### PR TITLE
DT-6352 start search from the user location when in the configured waltti area

### PR DESCRIPTION
**Pull request by sharhio**
## Proposed Changes

  - When user has already given the permission to use location, add current location as the starting point for a route search by default. 
  - Only do this if the user is within the configured service area of the instance. 
  - Configurable on/off in the instance settings, configured for all waltti instances.

## Pull Request Check List

  - A reasonable set of unit tests is included
  - Console does not show new warnings/errors
  - Changes are documented or they are self explanatory
  - This pull request does not have any merge conflicts
  - All existing tests pass in CI build

## Review

  - Read and verify the code changes
  - Test the functionality by running the UI locally with all popular browsers available in your platform
  - Check that the implementation matches the design, when such one is defined in a Jira issue
  - Merge the pull request

### CODE REVIEW

File: app/component/IndexPage.js [hunk 1]
```diff
@@ -105,6 +110,18 @@ class IndexPage extends React.Component {
       this.context.executeAction(storeDestination, destination);
     }
 
+    if (this.context.config.startSearchFromUserLocation) {
+      checkPositioningPermission().then(permission => {
+        if (
+          permission.state === 'granted' &&
+          this.props.locationState.hasLocation === false &&
+          !this.props.locationState.isLocationingInProgress &&
+          this.props.locationState.status === 'no-location'
```
Comment (vesameskanen) [inline app/component/IndexPage.js:117]: Complex if chains are sort of unpleasant. Would it be enough to check state === 'granted' && status === 'no-location' ?

File: app/component/IndexPage.js [hunk 2]
```diff
@@ -280,6 +302,17 @@ class IndexPage extends React.Component {
...
     ];
 
+    if (
+      config.startSearchFromUserLocation &&
+      currentLocation &&
+      config.areaPolygon
+    ) {
+      const originPoint = [currentLocation.lon, currentLocation.lat];
+      if (inside(originPoint, config.areaPolygon)) {
+        this.context.executeAction(storeOrigin, currentLocation);
```
Comment (vesameskanen) [inline app/component/IndexPage.js]: render should not have any side effects. Debugger gives warnings about cascading updates. Maybe this  code should be in componentDidUpdate.

**Review by vesameskanen — CHANGES_REQUESTED** [review]

**Comment by vesameskanen** [comment]
Other observations: I tested kuopio reittiopas. Origin was automatically placed to tampere. This is surprising.

File: app/component/IndexPage.js [hunk 3]
```diff
@@ -280,6 +302,17 @@ class IndexPage extends React.Component {
       'Stops',
     ];
 
+    if (
+      config.startSearchFromUserLocation &&
+      currentLocation &&
+      config.areaPolygon
```
Comment (vesameskanen) [inline app/component/IndexPage.js]: It is  not necessary to test areaPolygon. Bad config should crash the UI.

File: app/component/IndexPage.js (hunk 2 above)
Comment (sharhio) [reply app/component/IndexPage.js]: Absolutely! I think the strange Kuopio glitch was caused by the silly placement, I managed to get the same problem but only in Kuopio for some reason. It didn't happen anymore after I changed it.

File: app/component/IndexPage.js [hunk 4]
```diff
@@ -128,6 +143,19 @@ class IndexPage extends React.Component {
...
     const { location } = match;
 
+    const currentLocation =
+      config.startSearchFromUserLocation &&
+      !this.props.origin.address &&
+      this.props.locationState?.hasLocation &&
+      this.props.locationState;
+
+    if (config.startSearchFromUserLocation && currentLocation) {
+      const originPoint = [currentLocation.lon, currentLocation.lat];
```
Comment (vesameskanen) [inline app/component/IndexPage.js:152]: This code is now executed many times. I suggest checking if currentLocation.reverseGeocoding is no longer in progress and origin and currentLocation coords are different or something like that.

**Review by vesameskanen — CHANGES_REQUESTED** [review]

File: app/component/IndexPage.js [hunk 5]
```diff
@@ -148,8 +148,11 @@ class IndexPage extends React.Component {
       !this.props.origin.address &&
       this.props.locationState?.hasLocation &&
       this.props.locationState;
-
-    if (config.startSearchFromUserLocation && currentLocation) {
+    if (
+      config.startSearchFromUserLocation &&
+      currentLocation &&
```
Comment (vesameskanen) [inline app/component/IndexPage.js]: One minor thing more: no ned to check config.startSearchFromUserLocation again, it was alredy checked when setting currentLocation.

**Review by vesameskanen — CHANGES_REQUESTED** [review]

File: app/component/IndexPage.js (hunk 5 above)
Comment (sharhio) [reply app/component/IndexPage.js]: Oh no, so it was!

**Pull request merged**